cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/realtime/source/KafkaSource.java

```diff
@@ -101,12 +121,28 @@ public SourceState poll(Emitter<StructuredRecord> writer, SourceState currentSta
   }
 
   /**
-   * Convert {@code Apache Kafka} ByetBuffer from message into CDAP {@link StructuredRecord} instance.
+   * Convert {@code Apache Kafka} ByteBuffer from message into CDAP {@link StructuredRecord} instance.
    * @param key the String key of the Kafka message
    * @param payload the ByteBuffer of the Kafka message.
-   * @return instance of {@link StructuredRecord} representing the message.
+   * @return instance of {@link StructuredRecord} representing the message using the appropriate format.
    */
   public StructuredRecord byteBufferToStructuredRecord(@Nullable String key, ByteBuffer payload) {
+    FormatSpecification spec = config.getFormatSpec();
+    if (config.format == null) {
+      return byteBufferToByteRecord(key, payload);
+    }
+    RecordFormat<StreamEvent, StructuredRecord> format;
+    try {
+      format = RecordFormats.createInitializedFormat(spec);
+      StreamEvent toStream = new StreamEvent(payload);
+      return format.read(toStream);
+    } catch (Exception e) {
+      LOG.debug("Could not parse Kafka payload into schema. Using default structured record instead.");
+      return byteBufferToByteRecord(key, payload);
+    }
+  }
+
+  private StructuredRecord byteBufferToByteRecord(@Nullable String key, ByteBuffer payload) {
```

Comment:
The return value is StructuredRecord, I would reccommend to change it to something like byteBufferToSchemalessByteRecord